libbeat/processors/rate_limit/algorithm/token_bucket.go

```diff
@@ -0,0 +1,202 @@
+
+	// NumBuckets is the number of buckets being utilized by the token
+	// bucket algorithm. When more than the specified number are utilized,
+	// GC is performed.
+	NumBuckets uint `config:"num_buckets"`
+}
+
+type tokenBucketConfig struct {
+	BurstMultiplier float64 `config:"burst_multiplier"`
+
+	// GC governs when completely filled token buckets must be deleted
+	// to free up memory. GC is performed when _any_ of the GC conditions
+	// below are met. After each GC, counters corresponding to _each_ of
+	// the GC conditions below are reset.
+	GC tokenBucketGCConfig `config:"gc"`
+}
+
+func newTokenBucket(config Config) (Algorithm, error) {
+	cfg := tokenBucketConfig{
+		BurstMultiplier: 1.0,
+		GC: tokenBucketGCConfig{
+			NumCalls:   10000,
+			NumBuckets: 1000,
+		},
+	}
+
+	if err := config.Config.Unpack(&cfg); err != nil {
+		return nil, errors.Wrap(err, "could not unpack token_bucket algorithm configuration")
+	}
+
+	return &tokenBucket{
+		limit:   config.Limit,
+		depth:   config.Limit.value * cfg.BurstMultiplier,
+		buckets: sync.Map{},
+		gc: struct {
+			thresholds tokenBucketGCConfig
+			metrics    struct {
+				numCalls   atomic.Uint
+				numBuckets atomic.Uint
+			}
+		}{
+			thresholds: tokenBucketGCConfig{
+				NumCalls:   cfg.GC.NumCalls,
+				NumBuckets: cfg.GC.NumBuckets,
+			},
+		},
+		clock: clockwork.NewRealClock(),
+	}, nil
+}
+
+func (t *tokenBucket) IsAllowed(key uint64) bool {
+	t.runGC()
+
+	b := t.getBucket(key)
+	allowed := b.withdraw()
+
+	t.gc.metrics.numCalls.Inc()
+	return allowed
+}
+
+// SetClock allows test code to inject a fake clock
+func (t *tokenBucket) SetClock(c clockwork.Clock) {
+	t.clock = c
+}
+
+func (t *tokenBucket) getBucket(key uint64) *bucket {
+	v, exists := t.buckets.LoadOrStore(key, &bucket{
+		tokens:        t.depth,
+		lastReplenish: t.clock.Now(),
+	})
+	b := v.(*bucket)
+
+	if exists {
+		b.replenish(t.limit, t.clock)
+		return b
+	}
+
+	t.gc.metrics.numBuckets.Inc()
+	return b
+}
+
+func (b *bucket) withdraw() bool {
+	if b.tokens < 1 {
+		return false
+	}
+	b.tokens--
+	return true
+}
+
+func (b *bucket) replenish(rate Rate, clock clockwork.Clock) {
+	secsSinceLastReplenish := clock.Now().Sub(b.lastReplenish).Seconds()
+	tokensToReplenish := secsSinceLastReplenish * rate.valuePerSecond()
+
+	b.tokens += tokensToReplenish
+	b.lastReplenish = clock.Now()
+}
+
+func (t *tokenBucket) runGC() {
+	// Don't run GC if thresholds haven't been crossed.
+	if (t.gc.metrics.numBuckets.Load() < t.gc.thresholds.NumBuckets) &&
+		(t.gc.metrics.numCalls.Load() < t.gc.thresholds.NumCalls) {
+		return
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// Don't run GC if thresholds haven't been crossed.
+	// Check again in case another GC thread has run while this GC thread
+	// was waiting to acquire the lock and the other GC thread reset the
+	// metrics.
+	if (t.gc.metrics.numBuckets.Load() < t.gc.thresholds.NumBuckets) &&
+		(t.gc.metrics.numCalls.Load() < t.gc.thresholds.NumCalls) {
+		return
+	}
```

Comment:
This is going to block all goroutines when a GC needs to be triggered, and then all of them will try to do the GC.

In my comment about this lock https://github.com/elastic/beats/pull/22883#discussion_r538614860 I was suggesting to use a "non-blocking mutex" (maybe this was not the more clear term, sorry :grimacing:).

I meant something like this, with a `unison.Mutex`:
```
      if !t.mu.TryLock() {
          return
      }
      defer t.mu.Unlock()
```

Or like this, with an atomic bool CAS:
```
    if !t.gcRunning.CAS(false, true) {
        return
    }
    defer t.gcRunning.Store(false)
```

With this, if a GC needs to be done, it is done in a single goroutine, and the rest of them will try to get the lock while the thresholds are crossed, but won't get blocked there and can go on with their tasks. And we don't need to check if the conditions are met again after getting the lock.

Reply:
Btw GC could be done in a different goroutine so we don't even block the current one:
```
      if !t.mu.TryLock() {
          return
      }
      go func() {
          defer t.mu.Unlock()
          ...
      }()
```

Reply:
Ah, gotcha, thanks for clarifying. I understand now what you meant by non-blocking mutex - if the lock is already taken by another goroutine, don't wait to acquire the lock but instead just move on (return). I will implement this.

++ to doing the actual GC steps in a different goroutine. Will implement this as well.

Reply:
Implemented non-blocking mutex in 3d8c0bee9ca405c08a358444b97257dff7ce5864.
Implemented separate goroutine for actual GC steps in f0fd35abd645fe27c80542b58e79eb46ba449c75.